drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -625,26 +651,66 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
 	return 0;
 }
 
-int usb_dc_ep_read_continue(uint8_t ep)
+int usb_dc_ep_read_continue(const uint8_t ep)
 {
-	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
-	size_t bytes_received;
+	const struct udc_rpi_ep_state *const ep_state = udc_rpi_get_ep_state(ep);
+	bool arm_out_endpoint = false;
 
 	if (!ep_state || !USB_EP_DIR_IS_OUT(ep)) {
 		LOG_ERR("Not valid endpoint: %02x", ep);
 		return -EINVAL;
 	}
+	if (ep == USB_CONTROL_EP_OUT && state.setup_available) {
+		const struct usb_setup_packet *setup = (const void *)&usb_dpram->setup_packet;
+
+		LOG_DBG("EP0 setup (wLength=%d, is_to_device=%d)",
+			setup->wLength, usb_reqtype_is_to_device(setup));
+		if (setup->wLength != 0U) {
+			/* In the case of a control transfer, we want to prime the OUT endpoint
+			 * exactly once, to either:
+			 * 1) in the to_device case, to receive the data (only if wLength is not 0)
+			 * 2) in the to_host case, to receive a 0-length status stage transfer
+			 *    (only valid if wLength is not 0)
+			 * Note that when wLength = 0, the status stage transfer is always an IN
+			 * type so we don't need to consider that case.
+			 */
+			arm_out_endpoint = true;
+			state.control_transfer_bytes_transferred = 0;
+		}
+		state.setup_available = false;
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(ep);
+
+		LOG_DBG("EP %d received %d offset: %d", ep, bytes_received, ep_state->read_offset);
+		if (ep == USB_CONTROL_EP_OUT) {
+			const struct usb_setup_packet *setup =
+				(const void *)&usb_dpram->setup_packet;
+			if (usb_reqtype_is_to_device(setup)) {
+				if (state.control_transfer_bytes_transferred + ep_state->read_offset
+					< setup->wLength) {
+					/* If no more data in the buffer, but we're still waiting
+					 * for more, start a new read transaction.
+					 */
+					if (bytes_received == ep_state->read_offset) {
+						state.control_transfer_bytes_transferred +=
+							ep_state->read_offset;
+						arm_out_endpoint = true;
+					}
+				}
+			}
+		} else {
+			/* If no more data in the buffer, start a new read transaction. */
+			if (bytes_received == ep_state->read_offset) {
+				arm_out_endpoint = true;
+			}
+		}
+	}
```

Comment:
Most of the new code is specific to control OUT endpoint, please move it to
`static int usb_dc_control_ep_read_continue(void)` function.

Reply:
DONE, I used the same name as you suggested, but I had to add some arguments, so it's not just (void)